### PR TITLE
[ISSUE #32] Implementation of RocketMQ Catalog 

### PR DESCRIPTION
**Pull request by li-xiao-shuang**
fix : #32

### CODE REVIEW

**Review by ShannonDing — COMMENTED** [review]
better to check CI error first.

**Comment by li-xiao-shuang** [comment]
> better to check CI error first.

done

**Review by GOODBOY008 — COMMENTED** [review]
@li-xiao-shuang I have left some comment.

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+                builder.column(field.getName(), toDataType);
+            }
+        }
+        Schema schema = builder.build();
+        Map<String, String> options = new HashMap<>();
+        options.put(RocketMqCatalogConstant.CONNECTOR, RocketMqCatalogConstant.ROCKETMQ_CONNECTOR);
+        options.put(RocketMqCatalogConstant.TOPIC, topic);
+        options.put(
+                RocketMqCatalogConstant.NAME_SERVER_ADDRESS,
+                mqAdminExt == null ? "" : mqAdminExt.getNamesrvAddr());
```
[the remark this answers is not in the document]
Comment (li-xiao-shuang) [reply src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:215]: This is for testable purposes

**Review by GOODBOY008 — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -182,7 +182,7 @@ CREATE TABLE rocketmq_source (
 ) WITH (
   'connector' = 'rocketmq',
   'topic' = 'user_behavior',
```
Comment (duhenglucky) [inline README.md:184]: How about adding a _namespace_ parameter?

Comment (li-xiao-shuang) [reply README.md:184]: If it is added in the table, then the user can add it by himself, but did not see the use of rocketmq-flink

Comment (duhenglucky) [reply README.md:184]: > If it is added in the table, then the user can add it by himself, but did not see the use of rocketmq-flink

ok, leave it in the next PR

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** Expose a RocketMQ instance as a database catalog. */
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:79]:
```suggestion
/** A catalog implementation for RocketMQ. */
```

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+    public static final String DEFAULT_DB = "default";
+    public final String namesrvAddr;
+    private final String schemaRegistryUrl;
+    private DefaultMQAdminExt mqAdminExt;
+    private SchemaRegistryClient schemaRegistryClient;
+
+    public RocketMQCatalog(
+            String catalogName, String database, String namesrvAddr, String schemaRegistryUrl) {
+        super(catalogName, database);
+        this.namesrvAddr = namesrvAddr;
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:92]: Adds `checkArgument(!isNullOrWhitespaceOnly(namesrvAddr), "namesrvAddr cannot be null or empty");`

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+        this.schemaRegistryUrl = schemaRegistryUrl;
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:93]: Adds checkArgument(!isNullOrWhitespaceOnly(schemaRegistryUrl), "schemaRegistryUrl cannot be null or empty");

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+        }
+    }
+
+    @Override
+    public void close() throws CatalogException {
+        if (Objects.nonNull(mqAdminExt)) {
+            mqAdminExt.shutdown();
+        }
+        if (Objects.nonNull(schemaRegistryClient)) {
+            schemaRegistryClient = null;
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:126]: Could the `schemaRegistryClient` firstly close and then set to null?

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+        }
+        if (schemaRegistryClient == null) {
+            schemaRegistryClient = SchemaRegistryClientFactory.newClient(schemaRegistryUrl, null);
+        }
+    }
+
+    @Override
+    public void close() throws CatalogException {
+        if (Objects.nonNull(mqAdminExt)) {
+            mqAdminExt.shutdown();
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:123]: Why not set `mqAdminExt ` to null?

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+    public List<String> listTables(String databaseName)
+            throws DatabaseNotExistException, CatalogException {
+        if (!getDefaultDatabase().equals(databaseName)) {
+            throw new DatabaseNotExistException(getName(), databaseName);
+        }
+        try {
+            List<String> tables = schemaRegistryClient.getSubjectsByTenant("default", "default");
+            return tables;
+        } catch (Exception e) {
+            throw new CatalogException("Fail to get topics from schema registry client.", e);
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:175]:
```suggestion
            throw new CatalogException(String.format("Failed to get topics of namespace %s from schema registry client.", databaseName), e);
```

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 9]
```diff
@@ -0,0 +1,471 @@
...
+        }
+        String subject = tablePath.getObjectName();
+        try {
+            GetSchemaResponse getSchemaResponse = schemaRegistryClient.getSchemaBySubject(subject);
+            if (getSchemaResponse.getType() != SchemaType.AVRO) {
+                throw new CatalogException("Only support avro schema.");
+            }
+            return getCatalogTableForSchema(subject, getSchemaResponse);
+        } catch (Exception e) {
+            throw new CatalogException("Fail to get schema from schema registry client.", e);
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:193]:
```suggestion
            throw new CatalogException(String.format("Failed to get schema of table %s from schema registry client.", tablePath.getFullName(), e);
```

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 10]
```diff
@@ -0,0 +1,471 @@
...
+            return false;
+        }
+        String subject = tablePath.getObjectName();
+        try {
+            GetSchemaResponse getSchemaResponse = schemaRegistryClient.getSchemaBySubject(subject);
+            if (Objects.nonNull(getSchemaResponse)) {
+                return true;
+            }
+        } catch (Exception e) {
+            throw new CatalogException("Fail to get topics from schema registry client.", e);
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:232]:
```suggestion
            throw new CatalogException(String.format("Failed to get schema of table %s from schema registry client.", tablePath.getFullName(), e);
```

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 11]
```diff
@@ -0,0 +1,471 @@
...
+    }
+
+    @Override
+    public void dropTable(ObjectPath tablePath, boolean ignoreIfNotExists)
+            throws TableNotExistException, CatalogException {
+        throw new UnsupportedOperationException();
+    }
+
+    // ------------------------------------------------------------------------
+    // Unsupported catalog operations for RocketMQ
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:250]: Why not move interfaces like `createDatabase` under this comment?

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 12]
```diff
@@ -0,0 +1,471 @@
...
+    }
+
+    @Override
+    public void renameTable(ObjectPath tablePath, String newTableName, boolean ignoreIfNotExists)
+            throws TableNotExistException, TableAlreadyExistException, CatalogException {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public List<CatalogPartitionSpec> listPartitions(ObjectPath tablePath)
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:317]: Moves this interface before the comment ` Unsupported catalog operations for RocketMQ`.

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 13]
```diff
@@ -0,0 +1,471 @@
...
+            throw new RuntimeException(e);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        } catch (MQBrokerException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public List<CatalogPartitionSpec> listPartitions(
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:349]: Ditto.

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 14]
```diff
@@ -0,0 +1,471 @@
...
+
+    @Override
+    public List<CatalogPartitionSpec> listPartitionsByFilter(
+            ObjectPath tablePath, List<Expression> expressions)
+            throws TableNotExistException, TableNotPartitionedException, CatalogException {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CatalogPartition getPartition(ObjectPath tablePath, CatalogPartitionSpec partitionSpec)
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:363]: Adds the implementation of the `getPartition` interface.

**Review by SteNicholas — CHANGES_REQUESTED** [review]

**Comment by SteNicholas** [comment]
@li-xiao-shuang, @GOODBOY008, thanks for your contribution. I left the comments for the changes of this pull request. BTW, please squash the commits to one commit. PTAL.

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java (hunk 6 above)
Comment (GOODBOY008) [reply src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:119]: `schemaRegistryClient` has not supply `close` or `shutdown` method.

File: pom.xml [hunk 15]
```diff
@@ -38,6 +38,7 @@
         <commons-lang.version>2.6</commons-lang.version>
         <spotless.version>2.4.2</spotless.version>
 	    <jaxb-api.version>2.3.1</jaxb-api.version>
```
Comment (SteNicholas) [inline pom.xml:40]: Fix the indent of this line.

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 16]
```diff
@@ -0,0 +1,493 @@
...
+                                            new HashMap<String, String>(1) {
+                                                {
+                                                    String queueId =
+                                                            String.valueOf(
+                                                                    topicOffset.getQueueId());
+                                                    put("__queue_id__", queueId);
+                                                }
+                                            }))
+                    .collect(Collectors.toList());
+        } catch (RemotingException e) {
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:268]: Could you please optimize the catch logic?

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java [hunk 17]
```diff
@@ -0,0 +1,493 @@
...
+        return new CatalogPartitionImpl(partitionSpec.getPartitionSpec(), null);
+    }
+
+    @Override
+    public boolean partitionExists(ObjectPath tablePath, CatalogPartitionSpec partitionSpec)
+            throws CatalogException {
+        try {
+            List<CatalogPartitionSpec> catalogPartitionSpecs = listPartitions(tablePath);
+            return catalogPartitionSpecs.contains(partitionSpec);
+        } catch (TableNotExistException e) {
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalog.java:298]: Ditto.

File: src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalogFactory.java [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.apache.rocketmq.flink.catalog.RocketMQCatalogFactoryOptions.DEFAULT_DATABASE;
+import static org.apache.rocketmq.flink.catalog.RocketMQCatalogFactoryOptions.IDENTIFIER;
+import static org.apache.rocketmq.flink.catalog.RocketMQCatalogFactoryOptions.NAME_SERVER_ADDR;
+import static org.apache.rocketmq.flink.catalog.RocketMQCatalogFactoryOptions.SCHEMA_REGISTRY_BASE_URL;
+
+/** RocketMQ {@CatalogFactory}. */
```
Comment (SteNicholas) [inline src/main/java/org/apache/rocketmq/flink/catalog/RocketMQCatalogFactory.java:35]:
```suggestion
/** The {@CatalogFactory} implementation of RocketMQ. */
```

**Review by SteNicholas — CHANGES_REQUESTED** [review]
@li-xiao-shuang @GOODBOY008 , thanks for the changes. Could you please rebase the master branch? I have updated the `RocketMQSource` for the `transformContext`. Thanks.

**Review by SteNicholas — APPROVED** [review]
LGTM. Thanks for @li-xiao-shuang @GOODBOY008 contribution of RocketMQ catalog.

**Pull request merged**